### PR TITLE
Include Test case for Limits

### LINKED ISSUE #81: Add memory limits as a test
faasd now has memory limit setting and reading, could we add a test for that?

cc @LucasRoesler @nitishkumar71

### PR DESCRIPTION
**Pull request by nitishkumar71**
Signed-off-by: Nitishkumar Singh <nitishkumarsingh71@gmail.com>

Certifier should support validation of memory limit in faasd and openfaas
## Motivation and Context

- [x] I have raised an issue to propose this change **this is required**

Closes #81 
## How Has This Been Tested?

1. New test case has been included to test memory limit
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Commits:

- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] My commit message has a body and describes how this was tested and why it is required.
- [x] I have signed-off my commits with `git commit -s` for the Developer Certificate of Origin (DCO)

Code:

- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.

Docs:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by nitishkumar71** [comment]
Can we run the workflow again? The issue is same as #82 where mostly failure happens for logs.

**Review by LucasRoesler — APPROVED** [review]

File: tests/deploy_test.go [hunk 1]
```diff
@@ -216,8 +231,14 @@ func compareDeployAndStatus(deploy types.FunctionDeployment, status types.Functi
 		return fmt.Errorf("incorrect Secrets: %s", err)
 	}
 
-	if !reflect.DeepEqual(deploy.Limits, status.Limits) {
+	if config.ProviderName != faasdProviderName && !reflect.DeepEqual(deploy.Limits, status.Limits) {
 		return fmt.Errorf("got %v, expected Limits %v", status.Limits, deploy.Limits)
+	} else if config.ProviderName == faasdProviderName && deploy.Limits != nil && status.Limits != nil {
+		deployLimits := *deploy.Limits
+		statusLimits := *status.Limits
+		if deployLimits.Memory != statusLimits.Memory {
```
[the remark this answers is not in the document]
Comment (nitishkumar71) [reply tests/deploy_test.go]: Changes are done. Since faasd does not support cpu, special handling is required for faasd.

**Comment by alexellis** [comment]
You can run this locally with `go test -run "^NameOfTest$" ` and then see the individual test fail for faasd to show that the memory limit wasn't detected. Then we'll cut a new release of faasd we can re-run the tests and see it pass.

File: tests/deploy_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nitishkumar71) [reply tests/deploy_test.go]: Changes are done to allow current stable faasd version to fail for Limits

**Comment by alexellis** [comment]
> 2021-12-05T17:50:48.3391383Z     --- FAIL: Test_Deploy_MetaData/test_listing_functions (0.02s)
> 2021-12-05T17:50:48.3392022Z         deploy_test.go:162: got nil, expected Limits &{5M 100m}

Perfect!

**Comment by alexellis** [comment]
We now have a new release which populates the memory value in the API, so let's run these tests again once the binaries are ready.

After that, if the faasd tests pass, I'll merge this PR.

https://github.com/openfaas/faasd/releases/tag/0.14.4

**Comment by nitishkumar71** [comment]
> We now have a new release which populates the memory value in the API, so let's run these tests again once the binaries are ready.
> 
> After that, if the faasd tests pass, I'll merge this PR.
> 
> https://github.com/openfaas/faasd/releases/tag/0.14.4

Can we run it now? I don't have permission for same.

**Comment by nitishkumar71** [comment]
@alexellis @LucasRoesler I have removed faasd memory test for now and have kept it just for faas-netes. Will work on it once we have resolved faasd#228

File: tests/deploy_test.go [hunk 2]
```diff
@@ -92,6 +92,21 @@ func Test_Deploy_MetaData(t *testing.T) {
...
+			name: "Deploy with memory limit",
+			function: types.FunctionDeployment{
+				Image:       imagePath,
+				Service:     "memory-limit",
+				EnvProcess:  "env",
+				Annotations: &map[string]string{},
+				Labels:      &map[string]string{},
+				Namespace:   config.DefaultNamespace,
+				Limits: &types.FunctionResources{
+					Memory: "5M",
```
Comment (alexellis) [inline tests/deploy_test.go:105]: `5Mi`?

Comment (nitishkumar71) [reply tests/deploy_test.go]: Changes are done and it works for both faas-netes and faasd.

**Review by LucasRoesler — CHANGES_REQUESTED** [review]
@nitishkumar71 i added a comment about improving the readability of the test by using a new config flag instead of the provider name. 

Another approach could be to have a function `SupportsCPULimits(name string) bool` that takes the provider name and returns true if it is in the supported list. But either way, using the provider name in the control flow is going to be a bit fragile in the long term.

File: tests/deploy_test.go [hunk 3]
```diff
@@ -216,8 +231,15 @@ func compareDeployAndStatus(deploy types.FunctionDeployment, status types.Functi
...
-	if !reflect.DeepEqual(deploy.Limits, status.Limits) {
+	if config.ProviderName != faasdProviderName && !reflect.DeepEqual(deploy.Limits, status.Limits) {
 		return fmt.Errorf("got %v, expected Limits %v", status.Limits, deploy.Limits)
+	} else if config.ProviderName == faasdProviderName && deploy.Limits != nil {
+		if status.Limits == nil {
+			return fmt.Errorf("got nil, expected Limits %v", deploy.Limits)
+		}
+		if deploy.Limits.Memory != status.Limits.Memory {
+			return fmt.Errorf("got %s, expected Requested Limit %s", status.Limits.Memory, deploy.Limits.Memory)
+		}
```
[the remark this answers is not in the document]
Comment (nitishkumar71) [reply tests/deploy_test.go:242]: Thank you for suggestion. No doubt this is the best way to write the code. There is an issue in [faasd](https://github.com/openfaas/faasd/issues/232) due to which below lines will not work for functions for which no memory limit is explicitly set.

```
if deploy.Limits == nil && status.Limits != nil {
      return fmt.Errorf("got %s, expected nil", status.Limits)
}
```

[the remark this answers is not in the document]
Comment (LucasRoesler) [reply tests/deploy_test.go]: @nitishkumar71 now that openfaas/faasd#232 is merged, what do we need to do to merge this?

[the remark this answers is not in the document]
Comment (LucasRoesler) [reply tests/deploy_test.go]: I think you need to force push again to trigger the actions to re-run the tests

[the remark this answers is not in the document]
Comment (nitishkumar71) [reply tests/deploy_test.go]: Done

**Review by LucasRoesler — APPROVED** [review]

**Comment by LucasRoesler** [comment]
I am going to merge this because the test that is failing is the scaling test, which has been flaky and we are still investigating. The new limits test is fine

**Pull request merged**